mobius-extras/src/main/java/com/spotify/mobius/extras/ToggledEventSource.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2020 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius.extras;
+
+import static com.spotify.mobius.internal_util.Preconditions.checkNotNull;
+
+import com.spotify.mobius.EventSource;
+import com.spotify.mobius.disposables.Disposable;
+import com.spotify.mobius.functions.Consumer;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * An event source that forwards the emitted values of a target Event Source, but whose emissions
+ * are toggled on/off by a secondary Boolean-emitting Event Source
+ *
+ * @param <T> The Event class
+ */
+public final class ToggledEventSource<T> implements EventSource<T> {
+  @Nonnull private final EventSource<T> targetEventSource;
+  @Nonnull private final List<Consumer<T>> consumers = new ArrayList<>(1);
+  @Nullable private Disposable disposable = null;
+  private Boolean active;
+
+  /**
+   * Returns a new EventSource that re-emits events from the <code>targetEventSource</code>
+   * depending on the value emitted by the <code>togglingEventSource</code><br>
+   * When the state is set to not re-emit values, this Event Source unsubscribed from the <code>
+   * targetEventSource</code> completely, and re-subscribes when the state is set to emit values
+   * again.
+   *
+   * @param <E> The class of the events being re-emitted
+   * @param targetEventSource The Event Source whose events will be conditionally re-emitted
+   * @param togglingSource The Event Source whose emitted values control whether the <code>
+   *     targetEventSource</code>'s values are re-emitted (when togglingSource emits True) or
+   *     discarded (when goggling Source emits False)
+   * @param initialToggleState The initial state of whether to re-emit values or not. True to allow
+   *     re-emitting, false to stop it.
+   */
+  @Nonnull
+  public static <E> EventSource<E> from(
+      @Nonnull EventSource<E> targetEventSource,
+      @Nonnull EventSource<Boolean> togglingSource,
+      boolean initialToggleState) {
+    final ToggledEventSource<E> toggledEventSource =
+        new ToggledEventSource<>(targetEventSource, initialToggleState);
+    new TogglingSourceObserver(togglingSource, toggledEventSource);
```

Comment:
I don't understand this implementation - is the idea that when the wrapped event source gets garbage collected, then the TogglingSourceObserver should stop subscribing to it?  (as in, dispose its subscription). Can that even happen? Is it possible to write this in a more explicit way - if not, could there be some more documentation on how it works and why that solution was chosen?

Reply:
Kind of, yes. The idea was to not keep the `toggledEventSource` (which is returned to the user) alive unless there's something else that has a reference to it. That's how I expect stand-alone EventSource objects to exist - only while something else can reach them.

My reasoning is like this: since this object (the `TogglingSourceObserver`) subscribes to the `togglingSource` at all times, that means the `togglingSource` would keep this object in memory, which would then keep the `toggledEventSource` alive too (unless I make it WeakRef'd). That would not make sense since there's no way to reach the `toggledEventSource` in any way from the `togglingEventSource`. 
With the weak ref, at least I can do some memory cleanup the next time the `togglingEventSource` changes true/false state.

I agree I should try to add some documentation to the constructor to make it clearer, and I'll check to make sure it isn't causing any issues.

Reply:
I also updated the name of the class, plus added some extra documentation to it, even though its internal.